File: src/services/obd2/oilLifeModel.ts
import { OilLifeProfile } from '../../types/obd';
import { CIVIC_2013_SPECS } from './civicSpecs';

const STORAGE_KEY = 'civic_2013_oil_profile_v1';

export class OilLifeEngine {
  private profile: OilLifeProfile;
  private currentTripDurationSec: number = 0;
  private currentTripMaxTempC: number = 0;

  constructor() {
    this.profile = this.loadProfile();
  }

  private getDefaultProfile(): OilLifeProfile {
    return {
      lastResetTimestamp: Date.now() - 30 * 24 * 60 * 60 * 1000, // 30 days ago default
      lastResetOdometer: 112000,
      currentOdometer: 114250,
      oilLifePercent: 78.5,
      accumulatedRevolutions: 3200000,
      coldStartsCount: 42,
      timeBelowOperatingTempSec: 28400,
      shortTripsCount: 14,
      highThermalStressSec: 920,
      estimatedMilesRemaining: 5887,
      estimatedDaysRemaining: 74,
      oilConditionGrade: 'Good',
      degradationBreakdown: {
        revWearFactor: 22.0,
        coldStartPenalty: 4.8,
        shortTripPenalty: 3.2,
        thermalShearPenalty: 1.5,
      },
    };
  }

  public loadProfile(): OilLifeProfile {
    try {
      const saved = localStorage.getItem(STORAGE_KEY);
      if (saved) {
        return JSON.parse(saved);
      }
    } catch {
      // Fallback
    }
    const defaultProf = this.getDefaultProfile();
    this.saveProfile(defaultProf);
    return defaultProf;
  }

  public saveProfile(profile: OilLifeProfile = this.profile): void {
    this.profile = profile;
    try {
      localStorage.setItem(STORAGE_KEY, JSON.stringify(profile));
    } catch {
      // Ignore storage quota errors
    }
  }

  public getProfile(): OilLifeProfile {
    return { ...this.profile };
  }

  /**
   * Called on each live engine update to accumulate wear factors.
   * @param rpm Current engine speed
   * @param coolantTempC Current engine coolant temperature in Celsius
   * @param engineLoadPercent Calculated engine load (0 - 100%)
   * @param speedMph Current vehicle speed
   * @param dtSec Delta time since last update (seconds)
   */
  public recordTelemetryStep(
    rpm: number,
    coolantTempC: number,
    engineLoadPercent: number,
    speedMph: number,
    dtSec: number
  ): OilLifeProfile {
    if (rpm < 400) {
      // Engine is not running
      return this.profile;
    }

    this.currentTripDurationSec += dtSec;
    this.currentTripMaxTempC = Math.max(this.currentTripMaxTempC, coolantTempC);

    // 1. Mechanical Revolutions Accumulation
    const stepRevs = (rpm / 60) * dtSec;
    this.profile.accumulatedRevolutions += stepRevs;

    // Mileage increment estimation
    const stepMiles = (speedMph / 3600) * dtSec;
    this.profile.currentOdometer += stepMiles;

    // 2. Cold Operation & Warmup Penalty
    // Normal operating temperature is ~71°C to 90°C (160°F - 195°F)
    const isCold = coolantTempC < CIVIC_2013_SPECS.operatingTempThresholdC;
    if (isCold) {
      this.profile.timeBelowOperatingTempSec += dtSec;
    }

    // 3. High Thermal / RPM Stress Penalty
    if (rpm > CIVIC_2013_SPECS.highThermalThresholdRpm && engineLoadPercent > CIVIC_2013_SPECS.highLoadThresholdPercent) {
      this.profile.highThermalStressSec += dtSec;
    }

    // Recompute Remaining Oil Life %
    this.recalculateOilHealth();
    this.saveProfile();
    return { ...this.profile };
  }

  /**
   * Marks a new engine start. If coolant is below 60°C (140°F), increments cold start.
   */
  public registerEngineStart(coolantTempC: number): void {
    this.currentTripDurationSec = 0;
    this.currentTripMaxTempC = coolantTempC;

    if (coolantTempC < 60) {
      this.profile.coldStartsCount++;
    }
    this.saveProfile();
  }

  /**
   * Marks trip completion: detects short trips (< 15 mins without reaching 85°C)
   * which cause moisture and fuel dilution in engine oil.
   */
  public registerEngineStop(): void {
    if (this.currentTripDurationSec > 60 && this.currentTripDurationSec < 900 && this.currentTripMaxTempC < 80) {
      // Trip under 15 minutes that didn't get hot enough to vaporize fuel/water condensation
      this.profile.shortTripsCount++;
    }
    this.currentTripDurationSec = 0;
    this.currentTripMaxTempC = 0;
    this.recalculateOilHealth();
    this.saveProfile();
  }

  /**
   * Resets oil life tracker to 100% after an oil & filter change.
   */
  public resetOilLife(odometerAtReset?: number): OilLifeProfile {
    const currentOdo = odometerAtReset || this.profile.currentOdometer;
    this.profile = {
      lastResetTimestamp: Date.now(),
      lastResetOdometer: currentOdo,
      currentOdometer: currentOdo,
      oilLifePercent: 100.0,
      accumulatedRevolutions: 0,
      coldStartsCount: 0,
      timeBelowOperatingTempSec: 0,
      shortTripsCount: 0,
      highThermalStressSec: 0,
      estimatedMilesRemaining: CIVIC_2013_SPECS.baselineOilLifeMiles,
      estimatedDaysRemaining: 180, // ~6 months default
      oilConditionGrade: 'Excellent',
      degradationBreakdown: {
        revWearFactor: 0,
        coldStartPenalty: 0,
        shortTripPenalty: 0,
        thermalShearPenalty: 0,
      },
    };
    this.saveProfile();
    return { ...this.profile };
  }

  private recalculateOilHealth(): void {
    // 1. Baseline Revolutions Wear:
    // 14.5M revolutions = 100% baseline wear
    const revWearPercent = (this.profile.accumulatedRevolutions / CIVIC_2013_SPECS.baselineLifetimeRevolutions) * 100;

    // 2. Cold Start & Warmup Penalty (each cold start & minute below 160°F shears oil molecules)
    const coldStartPenaltyPercent = (this.profile.coldStartsCount * 0.15) + (this.profile.timeBelowOperatingTempSec / 3600) * 0.4;

    // 3. Short Trip Moisture Dilution Penalty
    const shortTripPenaltyPercent = this.profile.shortTripsCount * 0.35;

    // 4. High Thermal / RPM Shear Penalty
    const thermalPenaltyPercent = (this.profile.highThermalStressSec / 60) * 0.25;

    const totalDegradation = revWearPercent + coldStartPenaltyPercent + shortTripPenaltyPercent + thermalPenaltyPercent;
    const remainingPercent = Math.max(0, Math.min(100, 100.0 - totalDegradation));

    this.profile.oilLifePercent = parseFloat(remainingPercent.toFixed(1));
    this.profile.degradationBreakdown = {
      revWearFactor: parseFloat(revWearPercent.toFixed(1)),
      coldStartPenalty: parseFloat(coldStartPenaltyPercent.toFixed(1)),
      shortTripPenalty: parseFloat(shortTripPenaltyPercent.toFixed(1)),
      thermalShearPenalty: parseFloat(thermalPenaltyPercent.toFixed(1)),
    };

    // Estimated Miles Remaining
    const milesDriven = Math.max(0, this.profile.currentOdometer - this.profile.lastResetOdometer);
    const degradationRatio = totalDegradation > 0 ? totalDegradation / 100 : 0.01;
    const effectiveTotalMiles = milesDriven > 50 ? (milesDriven / degradationRatio) : CIVIC_2013_SPECS.baselineOilLifeMiles;
    this.profile.estimatedMilesRemaining = Math.max(0, Math.round((remainingPercent / 100) * effectiveTotalMiles));

    // Condition Grade
    if (remainingPercent > 70) {
      this.profile.oilConditionGrade = 'Excellent';
    } else if (remainingPercent > 40) {
      this.profile.oilConditionGrade = 'Good';
    } else if (remainingPercent > 15) {
      this.profile.oilConditionGrade = 'Fair';
    } else if (remainingPercent > 5) {
      this.profile.oilConditionGrade = 'Service Due';
    } else {
      this.profile.oilConditionGrade = 'Degraded';
    }
  }
}

File: app-native/core/src/main/kotlin/com/shieldrj/civic5mt/core/OilLifeModel.kt
package com.shieldrj.civic5mt.core

import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.roundToLong

enum class OilConditionGrade(val label: String) {
    EXCELLENT("Excellent"),
    GOOD("Good"),
    FAIR("Fair"),
    SERVICE_DUE("Service Due"),
    DEGRADED("Degraded"),
    ;

    override fun toString(): String = label
}

data class DegradationBreakdown(
    val revWearFactor: Double,
    val coldStartPenalty: Double,
    val shortTripPenalty: Double,
    val thermalShearPenalty: Double,
)

data class OilLifeProfile(
    val lastResetTimestamp: Long,
    val lastResetOdometer: Double,
    val currentOdometer: Double,
    /** 0 - 100%. */
    val oilLifePercent: Double,

    // Deep tracking factors
    /** Total crankshaft cycles. */
    val accumulatedRevolutions: Double,
    /** Starts with coolant below 160°F. */
    val coldStartsCount: Int,
    /** Seconds under 160°F. */
    val timeBelowOperatingTempSec: Double,
    /** Trips under 15 minutes without full warmup. */
    val shortTripsCount: Int,
    /** Seconds above 4500 RPM at high load. */
    val highThermalStressSec: Double,

    // Projected wear
    val estimatedMilesRemaining: Int,
    val estimatedDaysRemaining: Int,
    val oilConditionGrade: OilConditionGrade,
    val degradationBreakdown: DegradationBreakdown,
)

/**
 * Where the oil profile is kept between runs.
 *
 * An interface because this module has no Android in it and must not gain any: the real
 * implementation will be Room, the tests use [InMemoryOilProfileStore], and the migration
 * importer will hand over the values rescued from the WebView's localStorage. The model
 * itself does not care which.
 */
interface OilProfileStore {
    fun load(): OilLifeProfile?
    fun save(profile: OilLifeProfile)
}

class InMemoryOilProfileStore(private var stored: OilLifeProfile? = null) : OilProfileStore {
    override fun load(): OilLifeProfile? = stored
    override fun save(profile: OilLifeProfile) {
        stored = profile
    }
}

/** Round to one decimal, the way the TypeScript's `parseFloat(x.toFixed(1))` did. */
private fun round1(value: Double): Double = toFixed(value, 1).toDouble()

class OilLifeEngine(
    private val store: OilProfileStore = InMemoryOilProfileStore(),
    private val clock: MillisClock = SystemMillisClock,
) {
    private var profile: OilLifeProfile
    private var currentTripDurationSec: Double = 0.0
    private var currentTripMaxTempC: Double = 0.0
    private var lastSaveTimestamp: Long = 0L

    init {
        profile = store.load() ?: defaultProfile().also { store.save(it) }
    }

    private fun defaultProfile(): OilLifeProfile = OilLifeProfile(
        lastResetTimestamp = clock.nowMillis() - 30L * 24 * 60 * 60 * 1000, // 30 days ago
        lastResetOdometer = 112000.0,
        currentOdometer = 114250.0,
        oilLifePercent = 78.5,
        accumulatedRevolutions = 3_200_000.0,
        coldStartsCount = 42,
        timeBelowOperatingTempSec = 28400.0,
        shortTripsCount = 14,
        highThermalStressSec = 920.0,
        estimatedMilesRemaining = 5887,
        estimatedDaysRemaining = 74,
        oilConditionGrade = OilConditionGrade.GOOD,
        degradationBreakdown = DegradationBreakdown(
            revWearFactor = 22.0,
            coldStartPenalty = 4.8,
            shortTripPenalty = 3.2,
            thermalShearPenalty = 1.5,
        ),
    )

    fun getProfile(): OilLifeProfile = profile

    fun saveProfile(toSave: OilLifeProfile = profile) {
        profile = toSave
        store.save(toSave)
    }

    /** Save at most once per 30 seconds - this runs on every telemetry tick. */
    private fun debouncedSave() {
        val now = clock.nowMillis()
        if (now - lastSaveTimestamp >= 30_000) {
            lastSaveTimestamp = now
            saveProfile()
        }
    }

    /**
     * Accumulates wear factors from one live engine update.
     *
     * @param rpm current engine speed
     * @param coolantTempC coolant temperature in Celsius
     * @param engineLoadPercent calculated engine load, 0-100
     * @param speedMph vehicle speed
     * @param dtSec seconds since the last update
     */
    fun recordTelemetryStep(
        rpm: Double,
        coolantTempC: Double,
        engineLoadPercent: Double,
        speedMph: Double,
        dtSec: Double,
    ): OilLifeProfile {
        if (rpm < 400) {
            // Engine is not running.
            return profile
        }

        currentTripDurationSec += dtSec
        currentTripMaxTempC = max(currentTripMaxTempC, coolantTempC)

        // 1. Mechanical revolutions.
        val stepRevs = (rpm / 60.0) * dtSec
        // Mileage increment estimate.
        val stepMiles = (speedMph / 3600.0) * dtSec

        // 2. Cold operation and warmup penalty. Normal operating temperature is ~71-90°C.
        val isCold = coolantTempC < CivicSpecs.OPERATING_TEMP_THRESHOLD_C

        // 3. High thermal / RPM stress penalty.
        val isHighStress = rpm > CivicSpecs.HIGH_THERMAL_THRESHOLD_RPM &&
            engineLoadPercent > CivicSpecs.HIGH_LOAD_THRESHOLD_PERCENT

        profile = profile.copy(
            accumulatedRevolutions = profile.accumulatedRevolutions + stepRevs,
            currentOdometer = profile.currentOdometer + stepMiles,
            timeBelowOperatingTempSec = profile.timeBelowOperatingTempSec + if (isCold) dtSec else 0.0,
            highThermalStressSec = profile.highThermalStressSec + if (isHighStress) dtSec else 0.0,
        )

        recalculateOilHealth()
        debouncedSave()
        return profile
    }

    /**
     * Marks a new engine start. Below the operating-temperature threshold it counts as a
     * cold start.
     */
    fun registerEngineStart(coolantTempC: Double) {
        currentTripDurationSec = 0.0
        currentTripMaxTempC = coolantTempC

        if (coolantTempC < CivicSpecs.OPERATING_TEMP_THRESHOLD_C) {
            profile = profile.copy(coldStartsCount = profile.coldStartsCount + 1)
        }
        saveProfile()
    }

    /**
     * Marks trip completion. A trip under 15 minutes that never got hot enough to boil off
     * condensation leaves moisture and fuel in the oil, so it is counted against it.
     */
    fun registerEngineStop() {
        if (currentTripDurationSec > 60 && currentTripDurationSec < 900 && currentTripMaxTempC < 80) {
            profile = profile.copy(shortTripsCount = profile.shortTripsCount + 1)
        }
        currentTripDurationSec = 0.0
        currentTripMaxTempC = 0.0
        recalculateOilHealth()
        saveProfile()
    }

    /**
     * Resets the tracker to 100% after an oil and filter change.
     */
    fun resetOilLife(odometerAtReset: Double? = null): OilLifeProfile {
        val currentOdo = odometerAtReset ?: profile.currentOdometer
        profile = OilLifeProfile(
            lastResetTimestamp = clock.nowMillis(),
            lastResetOdometer = currentOdo,
            currentOdometer = currentOdo,
            oilLifePercent = 100.0,
            accumulatedRevolutions = 0.0,
            coldStartsCount = 0,
            timeBelowOperatingTempSec = 0.0,
            shortTripsCount = 0,
            highThermalStressSec = 0.0,
            estimatedMilesRemaining = CivicSpecs.BASELINE_OIL_LIFE_MILES.toInt(),
            estimatedDaysRemaining = 180, // ~6 months
            oilConditionGrade = OilConditionGrade.EXCELLENT,
            degradationBreakdown = DegradationBreakdown(0.0, 0.0, 0.0, 0.0),
        )
        saveProfile()
        return profile
    }

    private fun recalculateOilHealth() {
        // 1. Baseline revolutions wear: 14.5M revolutions = 100% baseline wear.
        val revWearPercent =
            (profile.accumulatedRevolutions / CivicSpecs.BASELINE_LIFETIME_REVOLUTIONS) * 100

        // 2. Cold start and warmup penalty - each cold start and each minute below 160°F
        //    shears oil molecules.
        val coldStartPenaltyPercent =
            (profile.coldStartsCount * 0.15) + (profile.timeBelowOperatingTempSec / 3600) * 0.4

        // 3. Short-trip moisture dilution penalty.
        val shortTripPenaltyPercent = profile.shortTripsCount * 0.35

        // 4. High thermal / RPM shear penalty.
        val thermalPenaltyPercent = (profile.highThermalStressSec / 60) * 0.25

        val totalDegradation =
            revWearPercent + coldStartPenaltyPercent + shortTripPenaltyPercent + thermalPenaltyPercent
        val remainingPercent = max(0.0, min(100.0, 100.0 - totalDegradation))

        // Estimated miles remaining.
        val milesDriven = max(0.0, profile.currentOdometer - profile.lastResetOdometer)
        val degradationRatio = if (totalDegradation > 0) totalDegradation / 100 else 0.01
        val effectiveTotalMiles =
            if (milesDriven > 50) milesDriven / degradationRatio else CivicSpecs.BASELINE_OIL_LIFE_MILES
        val estimatedMilesRemaining =
            max(0L, ((remainingPercent / 100) * effectiveTotalMiles).roundToLong()).toInt()

        // Days remaining, from the daily mileage burn rate.
        val daysSinceReset =
            max(1.0, (clock.nowMillis() - profile.lastResetTimestamp) / (24.0 * 60 * 60 * 1000))
        val dailyMileage = milesDriven / daysSinceReset
        val estimatedDaysRemaining =
            if (dailyMileage > 0.1) (estimatedMilesRemaining / dailyMileage).roundToInt() else 180

        val grade = when {
            remainingPercent > 70 -> OilConditionGrade.EXCELLENT
            remainingPercent > 40 -> OilConditionGrade.GOOD
            remainingPercent > 15 -> OilConditionGrade.FAIR
            remainingPercent > 5 -> OilConditionGrade.SERVICE_DUE
            else -> OilConditionGrade.DEGRADED
        }

        profile = profile.copy(
            oilLifePercent = round1(remainingPercent),
            degradationBreakdown = DegradationBreakdown(
                revWearFactor = round1(revWearPercent),
                coldStartPenalty = round1(coldStartPenaltyPercent),
                shortTripPenalty = round1(shortTripPenaltyPercent),
                thermalShearPenalty = round1(thermalPenaltyPercent),
            ),
            estimatedMilesRemaining = estimatedMilesRemaining,
            estimatedDaysRemaining = estimatedDaysRemaining,
            oilConditionGrade = grade,
        )
    }
}
